File: src/main/scala/com/vc/hard/L135.scala
package com.vc.hard

object L135 {
  def candy(arr: Array[Int]): Int = {
    val n = arr.length

    //Go from left to right
    var i = 1
    val left = Array.fill(n)(1)
    while(i < n) {
      if(arr(i) > arr(i - 1)) {
        left(i) = left(i - 1) + 1
      }
      i += 1
    }

    //Go from right to left
    i = n - 1
    val right = Array.fill(n)(1)
    while(i > 0) {
      if(arr(i - 1) > arr(i)) {
        right(i - 1) = right(i) + 1
      }
      i -= 1
    }

    //Take max at each position
    val total = new Array[Int](n)
    i = 0
    while(i < n) {
      total(i) = Math.max(left(i), right(i))
      i += 1
    }
    total.sum
  }
}

File: src/main/java/com/vc/hard/L135.java
package com.vc.hard;

import java.util.*;

class L135 {
    public int candy(int[] ratings) {
        int n = ratings.length;

        int[] left = new int[n];
        Arrays.fill(left, 1);
        int i = 1;
        while(i < n) {
            if(ratings[i - 1] < ratings[i]) {
                left[i] = left[i - 1] + 1;
            }
            i++;
        }

        int[] right = new int[n];
        Arrays.fill(right, 1);
        i = n - 1;
        while(i > 0) {
            if(ratings[i] < ratings[i - 1]) {
                right[i - 1] = right[i] + 1;
            }
            i--;
        }

        int[] total = new int[n];
        for(int j = 0; j < n; j++) {
            total[j] = Math.max(left[j], right[j]);
        }

        int res = 0;
        for(int j = 0; j < n; j++) {
            res += total[j];
        }
        return res;
    }
}
